oauth-server/src/main/java/io/strimzi/kafka/oauth/server/JaasServerOauthValidatorCallbackHandler.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright 2017-2019, Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.kafka.oauth.server;
+
+import io.strimzi.kafka.oauth.common.Config;
+import io.strimzi.kafka.oauth.common.ConfigUtil;
+import io.strimzi.kafka.oauth.validator.JWTSignatureValidator;
+import io.strimzi.kafka.oauth.validator.OAuthIntrospectionValidator;
+import io.strimzi.kafka.oauth.common.TokenInfo;
+import io.strimzi.kafka.oauth.validator.TokenValidator;
+import io.strimzi.kafka.oauth.validator.TokenValidationException;
+import org.apache.kafka.common.errors.AuthenticationException;
+import org.apache.kafka.common.security.auth.AuthenticateCallbackHandler;
+import org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule;
+import org.apache.kafka.common.security.oauthbearer.OAuthBearerToken;
+import org.apache.kafka.common.security.oauthbearer.OAuthBearerValidatorCallback;
+import org.keycloak.jose.jws.JWSInput;
+import org.keycloak.jose.jws.JWSInputException;
+import org.keycloak.representations.AccessToken;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.SSLSocketFactory;
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.auth.login.AppConfigurationEntry;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+import static io.strimzi.kafka.oauth.common.JSONUtil.getClaimFromJWT;
+import static io.strimzi.kafka.oauth.common.LogUtil.mask;
+
+public class JaasServerOauthValidatorCallbackHandler implements AuthenticateCallbackHandler {
+
+    private static final Logger log = LoggerFactory.getLogger(JaasServerOauthValidatorCallbackHandler.class);
+
+    private TokenValidator validator;
+
+    private ServerConfig config;
+
+    private String usernameClaim;
+
+    @Override
+    public void configure(Map<String, ?> configs, String saslMechanism, List<AppConfigurationEntry> jaasConfigEntries) {
+
+        if (!OAuthBearerLoginModule.OAUTHBEARER_MECHANISM.equals(saslMechanism))    {
+            throw new IllegalArgumentException(String.format("Unexpected SASL mechanism: %s", saslMechanism));
+        }
+
+        if (jaasConfigEntries.size() != 1) {
+            throw new IllegalArgumentException("Exactly one jaasConfigEntry expected (size: " + jaasConfigEntries.size());
+        }
+
+        for (AppConfigurationEntry e: jaasConfigEntries) {
+            Properties p = new Properties();
+            p.putAll(e.getOptions());
+            config = new ServerConfig(p);
+            break;
+        }
```

Comment:
It's not clear to me how this config works. You're taking some config from the JAAS config, but `ServerConfigs` also looks up in system properties and environment. And you're completely ignoring the `configs` parameter, which on the face of it is how I'd expect this to be configured. Can you explain, or refer me to an existing explanation (maybe I missed it?)

Reply:
I mostly reverse engineered this by stepping through code. Based on that configs seems to be a master configuration object of Kafka that gives you access to other configuration, possibly allowing you to inject your own things in there. Callback handler module specific configuration is passed via jaasConfigEntries.

ServerConfig constructor takes default values. That means you can override any value through system property or env variable. If not overridden that way use the one provided by jaasConfigEntries.
One TODO is to support per listener configurations - current code assumes there is one OAuth configuration.

Configuration.getValue() is the key to how System properties, env properties override mechanism works.

Reply:
To put it succinctly - the OAuth2 config uses its own system properties / environment variables / JAAS login module options. It makes no use of Kafka's server.properties.